hydra-node/test/Hydra/Model.hs

```diff
@@ -0,0 +1,579 @@
+              ]
+      party `performs` Input.Commit{Input.utxo = realUtxo}
+
+performNewTx ::
+  (MonadThrow m, MonadAsync m, MonadTimer m) =>
+  WorldState m ->
+  Party ->
+  Payment ->
+  StateT (Nodes m) m ()
+performNewTx st party tx = do
+  let recipient = mkVkAddress testNetworkId (getVerificationKey (to tx))
+  nodes <- gets nodes
+
+  let waitForOpen = do
+        outs <- lift $ serverOutputs (nodes ! party)
+        let matchHeadIsOpen = \case
+              Output.HeadIsOpen{} -> True
+              _ -> False
+        case find matchHeadIsOpen outs of
+          Nothing -> lift (threadDelay 0.1) >> waitForOpen
+          Just{} -> pure ()
+  waitForOpen
+
+  party `performs` Input.GetUTxO
+
+  let matchPayment p@(_, txOut) =
+        isOwned (from tx) p && value tx == txOutValue txOut
+
+      waitForUTxO utxo = \case
+        0 ->
+          lift $
+            throwIO $
+              Prelude.userError
+                ( "no utxo matched,\npayment = " <> show tx
+                    <> "\nutxo =  "
+                    <> show utxo
+                    <> "\nconfirmed = "
+                    <> show (fmap (first (mkVkAddress @Era testNetworkId . getVerificationKey)) $ confirmedUTxO $ offChainState $ hydraState st)
+                )
+        n ->
+          lift (threadDelay 1 >> waitForNext (nodes ! party)) >>= \case
+            GetUTxOResponse u
+              | u == mempty -> do
+                party `performs` Input.GetUTxO
+                waitForUTxO u (n -1)
+              | otherwise -> case find matchPayment (UTxO.pairs u) of
+                Nothing -> do
+                  party `performs` Input.GetUTxO
+                  waitForUTxO u (n -1)
+                Just p -> pure p
+            _ ->
+              waitForUTxO utxo (n -1)
+
+  (i, o) <- waitForUTxO mempty (5000 :: Int)
+
+  let realTx =
+        either
+          (error . show)
+          id
+          (mkSimpleTx (i, o) (recipient, value tx) (from tx))
+
+  party `performs` Input.NewTx realTx
+  lift $
+    waitUntilMatch [nodes ! party] $ \case
+      SnapshotConfirmed{Output.snapshot = snapshot} ->
+        realTx `elem` Snapshot.confirmed snapshot
+      err@Output.TxInvalid{} -> error ("expected tx to be valid: " <> show err)
+      _ -> False
+
+--
+
+-- * Generator Helpers
+
+--
+
+genPayment :: WorldState m -> Gen (Party, Payment)
+genPayment WorldState{hydraParties, hydraState} =
+  case hydraState of
+    Open{offChainState = OffChainState{confirmedUTxO}} -> do
+      (from, value) <-
+        elements confirmedUTxO `suchThat` (not . null . valueToList . snd)
+      let party = deriveParty $ fst $ fromJust $ List.find ((== from) . snd) hydraParties
+      (_, to) <- elements hydraParties `suchThat` ((/= from) . snd)
+      pure (party, Payment{from, to, value})
+    _ -> error $ "genPayment impossible in state: " <> show hydraState
+
+unsafeConstructorName :: (Show a) => a -> String
+unsafeConstructorName = Prelude.head . Prelude.words . show
```

Comment:
We could also use https://hackage.haskell.org/package/base-4.16.1.0/docs/Data-Data.html#v:toConstr and `show` that for a safe variant of this?

Reply:
This requires `Data` derivation which ripples all over the codebase. Not sure this is worth it.